continuous_integration/gpuci/build.sh

```diff
@@ -23,6 +23,9 @@ cd "$WORKSPACE"
 # Determine CUDA release version
 export CUDA_REL=${CUDA_VERSION%.*}
 
+# FIXME - monitoring GIL contention causes UCX teardown issues
+export DASK_DISTRIBUTED__ADMIN__SYSTEM_MONITOR__GIL__ENABLED=False
```

Comment:
FWIW I don't quite understand the underlying issue here. Maybe a separate, follow-up issue makes sense here

Reply:
I think the issue here is that the gilknocker was holding onto resources making the UCX clean difficult.  Disabling gilknocker explicitly during GPUCI tests unblocked ucx-py/ucx cleanup

Reply:
I guess I'm wondering what resource is gilknocker holding onto? Should gilknocker be updated to cleanup itself more, or should updates be made on the UCX side?

Reply:
Not immediately obvious to me - perhaps some relevant tracebacks here are that of the UCX teardown in the tests:

<details>

```python-traceback
_______________________________________________________________________________________________________________________________________ ERROR at teardown of test_simple ________________________________________________________________________________________________________________________________________

    @pytest.fixture(scope="function")
    def ucx_loop():
        """Allows UCX to cancel progress tasks before closing event loop.
    
        When UCX tasks are not completed in time (e.g., by unexpected Endpoint
        closure), clean up tasks before closing the event loop to prevent unwanted
        errors from being raised.
        """
        ucp = pytest.importorskip("ucp")
    
        loop = asyncio.new_event_loop()
        loop.set_exception_handler(ucx_exception_handler)
        ucp.reset()
        yield loop
>       ucp.reset()

distributed/utils_test.py:2148: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def reset():
        """Resets the UCX library by shutting down all of UCX.
    
        The library is initiated at next API call.
        """
        global _ctx
        if _ctx is not None:
            weakref_ctx = weakref.ref(_ctx)
            _ctx = None
            gc.collect()
            if weakref_ctx() is not None:
                msg = (
                    "Trying to reset UCX but not all Endpoints and/or Listeners "
                    "are closed(). The following objects are still referencing "
                    "ApplicationContext: "
                )
                for o in gc.get_referrers(weakref_ctx()):
                    msg += "\n  %s" % str(o)
>               raise UCXError(msg)
E               ucp._libs.exceptions.UCXError: Trying to reset UCX but not all Endpoints and/or Listeners are closed(). The following objects are still referencing ApplicationContext: 
E                 <frame at 0x5569f2351ac0, file '/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/site-packages/ucp/core.py', line 143, code _listener_handler_coroutine>

/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/site-packages/ucp/core.py:947: UCXError
--------------------------------------------------------------------------------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------------------------------------------------------------------------------
2023-05-01 11:05:44,980 - distributed.worker - ERROR - Failed to communicate with scheduler during heartbeat.
Traceback (most recent call last):
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1431, in _connect
    comm = await connect(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/comm/core.py", line 292, in connect
    comm = await wait_for(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils.py", line 1878, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/asyncio/tasks.py", line 432, in wait_for
    await waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nfs/charlesb/dev/distributed/main/distributed/worker.py", line 1237, in heartbeat
    response = await retry_operation(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils_comm.py", line 434, in retry_operation
    return await retry(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils_comm.py", line 413, in retry
    return await coro()
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1266, in send_recv_from_rpc
    comm = await self.pool.connect(self.addr)
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1510, in connect
    return await connect_attempt
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1454, in _connect
    raise CommClosedError(reason)
distributed.comm.core.CommClosedError: ConnectionPool closing.
2023-05-01 11:05:44,986 - distributed.worker - ERROR - Failed to communicate with scheduler during heartbeat.
Traceback (most recent call last):
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1431, in _connect
    comm = await connect(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/comm/core.py", line 292, in connect
    comm = await wait_for(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils.py", line 1878, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/asyncio/tasks.py", line 432, in wait_for
    await waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nfs/charlesb/dev/distributed/main/distributed/worker.py", line 1237, in heartbeat
    response = await retry_operation(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils_comm.py", line 434, in retry_operation
    return await retry(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils_comm.py", line 413, in retry
    return await coro()
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1266, in send_recv_from_rpc
    comm = await self.pool.connect(self.addr)
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1510, in connect
    return await connect_attempt
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1454, in _connect
    raise CommClosedError(reason)
distributed.comm.core.CommClosedError: ConnectionPool closing.
----------------------------------------------------------------------------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------------------------------------------------------------------------
ERROR    asyncio:base_events.py:1758 Task exception was never retrieved
future: <Task finished name='Task-356' coro=<_listener_handler_coroutine() done, defined at /datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/site-packages/ucp/core.py:128> exception=UCXError('<stream_recv>: Connection reset by remote peer')>
Traceback (most recent call last):
  File "/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/site-packages/ucp/core.py", line 143, in _listener_handler_coroutine
    peer_info = await exchange_peer_info(
  File "/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/site-packages/ucp/core.py", line 52, in exchange_peer_info
    await comm.stream_recv(endpoint, peer_info_arr, peer_info_arr.nbytes)
ucp._libs.exceptions.UCXError: <stream_recv>: Connection reset by remote peer
```

</details>

And a simple reproducer:

```python
from distributed import LocalCluster

cluster = LocalCluster(protocol="ucx")
cluster.close()
```

Which gives me the following worker errors:

```python-traceback
2023-05-01 12:46:05,381 - distributed.worker - ERROR - Failed to communicate with scheduler during heartbeat.
Traceback (most recent call last):
  File "/home/nfs/charlesb/dev/distributed/main/distributed/comm/ucx.py", line 349, in read
    await self.ep.recv(msg)
  File "/datasets/charlesb/mambaforge/envs/distributed-gpuci-py310/lib/python3.10/site-packages/ucp/core.py", line 725, in recv
    ret = await comm.tag_recv(self._ep, buffer, nbytes, tag, name=log)
ucp._libs.exceptions.UCXCanceled: <[Recv #183] ep: 0x7fd3f40730c0, tag: 0x42adb052da7b4792, nbytes: 16, type: <class 'numpy.ndarray'>>: 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nfs/charlesb/dev/distributed/main/distributed/worker.py", line 1237, in heartbeat
    response = await retry_operation(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils_comm.py", line 434, in retry_operation
    return await retry(
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils_comm.py", line 413, in retry
    return await coro()
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1269, in send_recv_from_rpc
    return await send_recv(comm=comm, op=key, **kwargs)
  File "/home/nfs/charlesb/dev/distributed/main/distributed/core.py", line 1028, in send_recv
    response = await comm.read(deserializers=deserializers)
  File "/home/nfs/charlesb/dev/distributed/main/distributed/utils.py", line 754, in wrapper
    return await func(*args, **kwargs)
  File "/home/nfs/charlesb/dev/distributed/main/distributed/comm/ucx.py", line 367, in read
    raise CommClosedError(
distributed.comm.core.CommClosedError: Connection closed by writer.
Inner exception: UCXCanceled("<[Recv #183] ep: 0x7fd3f40730c0, tag: 0x42adb052da7b4792, nbytes: 16, type: <class 'numpy.ndarray'>>: ")
```

Reply:
Thanks @charlesbluca, it's good to have that simple example. cc @milesgranger for visibility

Reply:
Thanks, I haven't seen that being a side-effect of gilknocker. In my experience, one can get "No Python frame" error which is a result of the monitoring thread attempting to acquire the GIL but the python process has already exited. I'm inclined to doubt it's the result of it holding onto anything as it [_immediately_](https://github.com/milesgranger/gilknocker/blob/70905f97c42b6a0e01f9d1fd56132b9c837a4734/src/lib.rs#L196) returns after acquiring the GIL. 

More likely, is the former, and to avoid that is to ensure `.stop()` is called before the process exits and that is [already done at a best effort](https://github.com/milesgranger/gilknocker/blob/70905f97c42b6a0e01f9d1fd56132b9c837a4734/src/lib.rs#L163) but won't capture anything which `atexit` wouldn't handle. 

What you're experiencing shouldn't happen anyway, gilknocker ought to stay out of the way, so this is unfortunate indeed.